### PR TITLE
[Bug Fix] wordpress/* - Fixed missing logs:CreateLogGroup permission

### PR DESCRIPTION
**Pull request by sriganesh**
…up failed with exception An error occurred (AccessDeniedException) when calling the CreateLogGroup operation: User: arn:aws:sts::XXXXXXXXXXXX:assumed-role/wordpress-WebServerIAMRole-XXXXXXXXXXXXX/i-XXXXXXXXXXXXXXXXX is not authorized to perform: logs:CreateLogGroup on resource: arn:aws:logs:us-east-1:XXXXXXXXXXXX:log-group:wordpress-WebServerLogs-XXXXXXXXXXXXX:log-stream: error because of missing permissions

**(Override all values in parentheses)**

(Run `yamllint folder/template.yaml`, `cfn-lint -i E1019 E3002 E2520 -t folder/template.yaml`, and `aws cloudformation validate-template --template-body file://folder/template.yaml` before you open a PR)

(Do not include multiple changes in one PR. Open additional PRs instead.)

---

(describe your change here)
Fixed error related to missing CreateLogGroup role permission while creating WebServerIAMRole

### CODE REVIEW

**Comment by michaelwittig** [comment]
The log group that is in use is already created by CloudFormation here: https://github.com/widdix/aws-cf-templates/blob/master/wordpress/wordpress-ha.yaml#L215

Where do you get this error from?

**Comment by sriganesh** [comment]
I am seeing the error continuously in /var/log/awslogs.log till I manually added the permission in the inline policy for the WebServerIAMRole.

**Comment by lwgordonimos** [comment]
FYI, I'm seeing this when deploying the Jenkins template as well:

```
    "eventTime": "2019-05-15T05:03:07Z",
    "eventSource": "logs.amazonaws.com",
    "eventName": "CreateLogGroup",
    "awsRegion": "ap-southeast-2",
    "sourceIPAddress": "XXX",
    "userAgent": "aws-cli/1.16.102 Python/2.7.14 Linux/4.14.104-95.84.amzn2.x86_64 botocore/1.12.92cwlogs/1.4.4",
    "errorCode": "AccessDenied",
    "errorMessage": "User: arn:aws:sts::XXX:assumed-role/jenkins-Jenkins-MasterIAMRole-XXX/i-XXX is not authorized to perform: logs:CreateLogGroup on resource: arn:aws:logs:ap-southeast-2:XXX:log-group:jenkins-Jenkins-MasterLogs-XXX:log-stream:",
    "requestParameters": null,
    "responseElements": null,
    "requestID": "ba62d37a-76ce-11e9-ab65-19a86d20e782",
    "eventID": "1258c538-fe40-495b-abcd-bb75ea09756c",
    "eventType": "AwsApiCall",
    "recipientAccountId": "XXX"
```

**Comment by michaelwittig** [comment]
@sriganesh which version of the wordpress template are you running  (check the TemplateVersion output of the stack)? can you confirm that logs are wirrten to CloudWatch Logs without your change?

@lwgordonimos which version of the jenkins template are you using (check the TemplateVersion output of the stack)? can you confirm that logs are wirrten to CloudWatch Logs?

**Comment by lwgordonimos** [comment]
@michaelwittig I've got to the bottom of it!

TL;DR: it only occurs when the SystemsManagerAccess template parameter is false because AmazonEC2RoleforSSM sneakily provides the `logs:CreateLogGroup` permission which *is* actually mandatory to use cwlogs.

Long version:

Despite the AWS doco at https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/AgentReference.html stating:

**I don't want the CloudWatch Logs agent to create either log groups or log streams automatically. How can I prevent the agent from recreating both log groups and log streams?**
*In your IAM policy, you can restrict the agent to only the following operations: DescribeLogStreams, PutLogEvents.*

This code seems to be executed when trying to push logs, and since the AccessDenied exception is not handled, it gets stuck in the retry loop and eventually just fails (apologies for the paste, I couldn't find the source code for this online):

Package: awscli-cwlogs
File: /usr/lib/python2.7/site-packages/cwlogs/push.py
```
  def _setup_resources(self):
        batch_logger.info('Creating log group %s.', self.log_group_name)
        resource_already_exists_ex = 'ResourceAlreadyExistsException'
        try:
            self.logs_service.create_log_group(
                logGroupName=self.log_group_name)
        except:
            type, value, traceback = exc_info()
            if not re.search(resource_already_exists_ex, str(value)):
                batch_logger.warning("CreateLogGroup failed with exception %s" % value)
                raise
        batch_logger.info('Creating log stream %s.', self.log_stream_name)
        try:
            self.logs_service.create_log_stream(
                logGroupName=self.log_group_name,
                logStreamName=self.log_stream_name)
        except:
            type, value, traceback = exc_info()
            if not re.search(resource_already_exists_ex, str(value)):
                batch_logger.warning("CreateLogStream failed with exception %s" % value)
                raise
```

```
2019-05-16 12:11:08,444 - cwlogs.push.publisher - WARNING - 9850 - Thread-857 - Caught exception: An error occurred (ResourceNotFoundException) when calling the PutLogEvents operation: The specified log stream does not exist.
2019-05-16 12:11:08,445 - cwlogs.push.batch - INFO - 9850 - Thread-857 - Creating log group LOG-GROUP-NAME.
2019-05-16 12:11:08,472 - cwlogs.push.batch - WARNING - 9850 - Thread-857 - CreateLogGroup failed with exception An error occurred (AccessDeniedException) when calling the CreateLogGroup operation: User: arn:aws:sts::ACCOUNT-ID:assumed-role/MASTER-IAM-ROLE/MASTER-INSTANCE-ID is not authorized to perform: logs:CreateLogGroup on resource: arn:aws:logs:ap-southeast-2:ACCOUNT-ID:log-group:LOG-GROUP-NAME:log-stream:
2019-05-16 12:11:08,473 - cwlogs.push.batch - WARNING - 9850 - Thread-857 - An error occurred (AccessDeniedException) when calling the CreateLogGroup operation: User: arn:aws:sts::ACCOUNT-ID:assumed-role/MASTER-IAM-ROLE/MASTER-INSTANCE-ID is not authorized to perform: logs:CreateLogGroup on resource: arn:aws:logs:ap-southeast-2:ACCOUNT-ID:log-group:LOG-GROUP-NAME:log-stream:
2019-05-16 12:11:08,473 - cwlogs.push.batch - WARNING - 9850 - Thread-857 - Method "_setup_resources" failed, backing off 0.50657440722 seconds, and retrying
2019-05-16 12:15:21,941 - cwlogs.threads - ERROR - 9850 - Thread-943 - Exception caught in <EventBatchPublisher(Thread-943, started daemon 140243743262464)>
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/cwlogs/threads.py", line 58, in run
    self._run()
  File "/usr/lib/python2.7/site-packages/cwlogs/push.py", line 1403, in _run
    self._publish_event_batch()
  File "/usr/lib/python2.7/site-packages/cwlogs/push.py", line 1210, in _publish_event_batch
    self.sequence_token = self._put_log_events(self.event_batch)
  File "/usr/lib/python2.7/site-packages/cwlogs/push.py", line 1294, in _put_log_events
    self._setup_resources
  File "/usr/lib/python2.7/site-packages/cwlogs/retry.py", line 116, in wrapper
    'attempts.' % (f.__name__, retry))
RuntimeError: Method "_setup_resources" has failed after 5 unsuccessful attempts.
```

This _setup_resources method being written like this means that the `logs:CreateLogGroup` is mandatory to even *test* whether the log group (and subsequently the log stream) already exists.

I suspect that since the SystemsManagerAccess parameter defaults to true, it's basically hiding this requirement, and that despite your noble efforts to run everything with least privilege, the cwlogs Python module makes this currently an impossibilty!

**Comment by lwgordonimos** [comment]
@sriganesh out of interest, did you also happen to turn off SystemsManagerAccess when deploying the stack?

**Comment by sriganesh** [comment]
@michaelwittig TemplateVersion | 9.11.0

@lwgordonimos Yes. I did set WebServerSystemsManagerAccess to false.

**Comment by michaelwittig** [comment]
@lwgordonimos good catch. I will merge soon and fix in all our other templates as well
